### PR TITLE
Refactor editor icon retrieval

### LINKED ISSUE #21615: Area2d Black Icon on Mac OSX 10.13
**Godot version:**

3.1 custom build

**OS/device including version:**

Mac Book  pro. Nvidia GT 750M

**Issue description:**

Area2d has black Icons in a scene
**Steps to reproduce:**

**Minimal reproduction project:**

add area2d as a child of a position2d with both as children of a kinematic2d
screen shot 2018-08-30 at 10 19 49 am

screen shot 2018-08-30 at 10 23 47 am

### PR DESCRIPTION
**Pull request by willnationsdev**
It looks like this has a ton of different changes, but the vast majority of this is just about centralizing copy-pasta code all over the Godot Editor into a centralized location that does the same processing in every case (rather than bits and pieces that the copy-pasta had induced at various locations).

Fixes #21615 and circumvents any previous "fixes" surrounding "_editor_icon" metadata.

1. Create `EditorData::script_class_get_name(p_path)` so that a path alone can fetch us the name, and from there, the icon path. Start caching these values during `EditorFileSystem::_scan_script_classes(...)`.
2. Create `EditorData::get_object_icon(obj, ...)` and `EditorData::get_object_icon_name(name, ...)` to fetch the appropriate editor icon depending on whether an object pointer or a String/StringName is available. Edit: we changed `get_object_icon_name` to `get_class_icon`.
3. Fix a script class icon bug in create_dialog.cpp regarding the `current_item_is_preferred` in the `cpp_type` case (prevents a possible `ClassDB::is_parent_class()` ERR_FAIL_COND_V trigger).
3. Replace all "_editor_icon" metadata fetches with the above functions. Allow editor nodes to check the live data directly rather than using the third-party "_editor_icon" metadata that was desynchronizing and causing all sorts of random bugs in different places.
    - group_editor.cpp
    - canvas_item_editor_plugin.cpp
    - spatial_editor_plugin.cpp
    - scene_tree_dock.cpp
    - scene_tree_editor.cpp
4. I then noticed that the same methods could be used to refactor and centralize the vast majority of `get_icon(...)` calls in the editor classes, so I started to evaluate all of those files and determine if their usage could be converted into an EditorData call without compromising their logical functionality. I believe I've done that, but I'm sure people will want to investigate it further.

### CODE REVIEW

**Comment by akien-mga** [comment]
There are some style issues: https://travis-ci.org/godotengine/godot/jobs/423784992

**Comment by willnationsdev** [comment]
@akien-mga Should be good to go, if someone wants to review it.

File: editor/animation_bezier_editor.cpp [hunk 1]
```diff
@@ -29,6 +29,7 @@
 /*************************************************************************/
 
 #include "animation_bezier_editor.h"
+#include "editor/editor_node.h"
```
Comment (akien-mga) [inline editor/animation_bezier_editor.cpp]: Please always leave an empty line between the .cpp's own header (`animation_bezier_editor.h`) and the other Godot headers.

**Comment by akien-mga** [comment]
Rebase needed after my change of `core` includes to be absolute.

File: editor/editor_data.cpp [hunk 2]
```diff
@@ -927,9 +951,87 @@ void EditorData::script_class_load_icon_paths() {
...
-		_script_class_icon_paths[key] = d[key];
+		String name = E->get().operator String();
+		_script_class_icon_paths[name] = d[name];
+
+		String path = ScriptServer::get_global_class_path(name);
+		script_class_set_name(path, name);
+	}
+}
+
+Ref<Texture> EditorData::get_object_icon(const Object *p_object, const Control *p_control, const String &p_fallback) const {
```
Comment (reduz) [inline editor/editor_data.cpp]: This is way too complicated. Make the function part of EditorNode and get the icon from the theme or from the gui_base instance, so you dont have to pass instances around.

**Review by reduz — DISMISSED** [review]

File: editor/editor_data.cpp (hunk 2 above)
Comment (willnationsdev) [reply editor/editor_data.cpp]: Done.

File: editor/editor_node.cpp [hunk 3]
```diff
@@ -3126,6 +3121,86 @@ void EditorNode::stop_child_process() {
...
+		}
+	}
+
+	if (p_fallback.length())
+		return gui_base->get_icon(p_fallback, "EditorIcons");
+
+	return NULL;
+}
+
+Ref<Texture> EditorNode::get_object_icon_name(const String &p_class, const String &p_fallback) const {
```
Comment (reduz) [inline editor/editor_node.cpp]: Two suggestions about this function:
* rename to get_class_icon (i think it makes more sense)
* fallback could have a default argument "Object" so it does not need to be passed always

**Review by reduz — DISMISSED** [review]

File: editor/editor_node.cpp (hunk 3 above)
Comment (willnationsdev) [reply editor/editor_node.cpp]: Done.

**Pull request merged**